core/vm/environment.go

```diff
@@ -17,110 +17,288 @@
 package vm
 
 import (
+	"fmt"
 	"math/big"
 
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/params"
 )
 
-// Environment is an EVM requirement and helper which allows access to outside
-// information such as states.
-type Environment interface {
-	// The current ruleset
-	ChainConfig() *params.ChainConfig
-	// The state database
-	Db() Database
-	// Creates a restorable snapshot
-	SnapshotDatabase() int
-	// Set database to previous snapshot
-	RevertToSnapshot(int)
-	// Address of the original invoker (first occurrence of the VM invoker)
-	Origin() common.Address
-	// The block number this VM is invoked on
-	BlockNumber() *big.Int
-	// The n'th hash ago from this block number
-	GetHash(uint64) common.Hash
-	// The handler's address
-	Coinbase() common.Address
-	// The current time (block time)
-	Time() *big.Int
-	// Difficulty set on the current block
-	Difficulty() *big.Int
-	// The gas limit of the block
-	GasLimit() *big.Int
-	// Determines whether it's possible to transact
-	CanTransfer(from common.Address, balance *big.Int) bool
-	// Transfers amount from one account to the other
-	Transfer(from, to Account, amount *big.Int)
-	// Adds a LOG to the state
-	AddLog(*Log)
-	// Type of the VM
-	Vm() Vm
-	// Get the curret calling depth
-	Depth() int
-	// Set the current calling depth
-	SetDepth(i int)
-	// Call another contract
-	Call(me ContractRef, addr common.Address, data []byte, gas, price, value *big.Int) ([]byte, error)
-	// Take another's contract code and execute within our own context
-	CallCode(me ContractRef, addr common.Address, data []byte, gas, price, value *big.Int) ([]byte, error)
-	// Same as CallCode except sender and value is propagated from parent to child scope
-	DelegateCall(me ContractRef, addr common.Address, data []byte, gas, price *big.Int) ([]byte, error)
-	// Create a new contract
-	Create(me ContractRef, data []byte, gas, price, value *big.Int) ([]byte, common.Address, error)
+type (
+	CanTransferFunc func(StateDB, common.Address, *big.Int) bool
```

Comment:
Throughout the code we used only `Fn` suffix to designate some callback type. Either `Fn` or `Func` works for me, but it would be nice to have it consistently the same throughout.